titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/jobmanager/JobCache.java

```diff
@@ -0,0 +1,251 @@
+package com.netflix.titus.runtime.connector.jobmanager;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.google.common.base.Preconditions;
+import com.netflix.titus.api.jobmanager.model.job.Job;
+import com.netflix.titus.api.jobmanager.model.job.JobState;
+import com.netflix.titus.api.jobmanager.model.job.Task;
+import com.netflix.titus.api.jobmanager.model.job.TaskState;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.tuple.Pair;
+
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+
+public class JobCache {
```

Comment:
As discussed. We should rename these types of classes to `Snapshot` since they are just an immutable snapshot of the state. Also, should start thinking about commonalities between the snapshots that should be followed either by setting a pattern or through code reuse:
* metrics
* data loading
* viewing the contents of a snapshot

Reply:
I will rename JobCache to snapshot in another PR. Too many changes are accumulated here.
This is still work in progress, and I purposefully did not create common APIs that could be used between GRPC clients (job, agent). Building a common abstraction for `Snapshot` reusable in different places is a larger piece of work. We can try to standardize some aspects (naming, metrics), but I do not believe we can build something as generic as Guava cache.